kafe2/fit/_base/cost.py

```diff
@@ -290,57 +225,94 @@ def on_no_errors(self):
-            else:
-                _chi2_func = self.chi2_pointwise_errors
+            _chi2_func = self.chi2_pointwise_errors
+            _arg_names = [self._DATA_NAME, self._MODEL_NAME, self._ERROR_NAME]
+            self._fail_on_no_matrix = False
+            self._fail_on_no_errors = not fallback_on_singular
             _cost_function_description += " (with pointwise errors)"
         else:
-            raise CostFunctionException(
-                "Unknown value '%s' for 'errors_to_use': must be one of ('covariance', 'pointwise', None)")
+            raise ValueError(
+                "Unknown value '%s' for 'errors_to_use': must be one of "
+                "('covariance', 'pointwise', None)")
 
-        super(CostFunctionBase_Chi2, self).__init__(cost_function=_chi2_func)
+        super(CostFunction_Chi2, self).__init__(
+            cost_function=_chi2_func,
+            arg_names=_arg_names)
 
         self._formatter.latex_name = "\\chi^2"
         self._formatter.name = "chi2"
         self._formatter.description = _cost_function_description
         self._needs_errors = _chi2_func is not self.chi2_no_errors
         self._is_chi2 = True
 
+    def _chi2(self, data, model, cov_mat_inverse=None, err=None):
+        data = np.asarray(data)
+        model = np.asarray(model)
+
+        if model.shape != data.shape:
+            raise ValueError(
+                "'data' and 'model' must have the same shape! Got %r and %r..."
+                % (data.shape, model.shape))
+
+        _res = (data - model)
+
+        # if a covariance matrix inverse is given, use it
+        if cov_mat_inverse is not None:
+            _cost = _res.dot(cov_mat_inverse).dot(_res)
+            if np.isnan(_cost):
+                _cost = np.inf
+            return _cost
+
+        if self._fail_on_no_matrix:
+            raise CostFunctionException("Covariance matrix is singular!")
+
+        # otherwise, if an array of point-wise errors is given, use that
+        if err is not None:
+            err = np.asarray(err)
+            if np.any(err == 0.0):
+                if self._fail_on_no_errors:
+                    raise CostFunctionException("'err' must not contain any zero values!")
+            else:
+                _res = _res / err
+
+        # return sum of squared residuals
+        _cost = np.sum(_res ** 2)
+        if np.isnan(_cost):
+            _cost = np.inf
+        return _cost
+
     def on_no_errors(self):
         if not self._no_errors_warning_printed:
-            if (self._cost_function_handle is self.chi2_covariance_fallback
-                    or self._cost_function_handle is self.chi2_pointwise_errors_fallback):
-                print('WARNING: No data errors were specified. Setting data errors to 1.')
-            else:
-                print('WARNING: No data errors were specified. The fit results may be wrong.')
+            print("WARNING: No data errors were specified. Setting data errors to 1.")
```

Comment:
This should be a warnings.warn() See https://docs.python.org/3/library/warnings.html